File: webclient/src/schema/service-job.ts
import { z } from "zod";
import { ServiceJobStatus } from "@/types/mia";
import { MIN_NOTES_LENGTH, MIN_SERVICE_INFO_LENGTH } from "@/constants/notesFieldConstants";

export const serviceJobBaseSchema = z.object({
  customer_id: z.string().min(1, "Customer is required"),
  vehicle_id: z.string().min(1, "Vehicle is required"),
  mechanic_id: z.string().min(1, "Mechanic is required"),
  service_info: z.string().min(MIN_SERVICE_INFO_LENGTH, `Service info must be at least ${MIN_SERVICE_INFO_LENGTH} characters`),
  mechanic_notes: z.string().min(MIN_NOTES_LENGTH, `Mechanic notes must be at least ${MIN_NOTES_LENGTH} characters`),
});

export const createServiceJobSchema = serviceJobBaseSchema.extend({
  id: z.string().min(1, "ID is required"),
  status: z.nativeEnum(ServiceJobStatus).default(ServiceJobStatus.Pending),
  validated_at: z.string().optional(),
  created_at: z.string().optional(),
});

export const updateServiceJobSchema = serviceJobBaseSchema.partial().extend({
  status: z.nativeEnum(ServiceJobStatus).optional(),
  validated_at: z.string().optional(),
});

export type ServiceJobFormValues = z.infer<typeof serviceJobBaseSchema>;
export type CreateServiceJobInput = z.infer<typeof createServiceJobSchema>;
export type UpdateServiceJobInput = z.infer<typeof updateServiceJobSchema>;

File: server/app/model/request/service_job.py
from pydantic import BaseModel, field_validator
from typing import Optional
from app.constants.enums.mia_enums import ServiceJobStatus
from datetime import datetime


class ServiceJobRequest(BaseModel):
    id: str
    customer_id: str
    vehicle_id: str
    mechanic_id: str
    status: ServiceJobStatus = ServiceJobStatus.pending
    service_info: str
    mechanic_notes: str
    validated_at: Optional[datetime] = None
    created_at: Optional[datetime] = None

    @field_validator("created_at", mode="before")
    @classmethod
    def set_created_at(cls, v):
        """Default to current time if created_at is None or not provided."""
        return v or datetime.utcnow()